gordon_janitor_gcp/gcp_clients.py

```diff
@@ -0,0 +1,234 @@
+
+    def __init__(self,
+                 auth_client=None,
+                 session=None,
+                 blacklisted_tags=None,
+                 blacklisted_metadata=None):
+        super().__init__(auth_client=auth_client, session=session)
+        self.blacklisted_tags = blacklisted_tags or []
+        self.blacklisted_metadata = blacklisted_metadata or []
+
+    async def list_instances(self,
+                             project,
+                             api_version='v1',
+                             page_size=500,
+                             instance_filter=None):
+        """Fetch all instances in a GCE project.
+
+        You can find the endpoint documentation
+        `here <https://cloud.google.com/compute/docs/reference/latest/
+        instances/aggregatedList>`__.
+
+        Args:
+            project (str): unique, user-provided project ID.
+            page_size (int): hint for the client to only retrieve up to this
+                number of results per API call.
+            instance_filter (str): endpoint-specific filter string used to
+                retrieve a subset of instances. This is passed directly to the
+                endpoint's "filter" URL query parameter.
+        Returns:
+            list: dicts containing instance data.
+        """
+        url = (f'{self.BASE_URL}{api_version}/projects/{project}'
+               '/aggregated/instances')
+        params = {'maxResults': page_size}
+        if instance_filter:
+            params['filter'] = instance_filter
+
+        responses = await self.list_all(url, params)
+        instances = self._parse_rsps_for_instances(responses)
+        return instances
+
+    def _parse_rsps_for_instances(self, responses):
+        instances = []
+        for response in responses:
+            for zone in response.get('items', {}).values():
+                instances.extend(self._filter_zone_instances(zone))
+        return instances
+
+    def _filter_zone_instances(self, zone):
+        instances = []
+        for instance in zone.get('instances', []):
+            if not any([
+                self._blacklisted_by_tag(instance),
+                self._blacklisted_by_metadata(instance)
+            ]):
+                try:
+                    instances.append(self._extract_instance_data(instance))
+                except (KeyError, IndexError) as e:
+                    logging.debug(
+                        'Could not extract instance information for '
+                        f'{instance} because of missing key {e}, skipping.')
+        return instances
+
+    def _extract_instance_data(self, instance):
+        iface_data = instance['networkInterfaces'][0]
+        return {
+            'hostname': instance['name'],
+            'internal_ip': iface_data['networkIP'],
+            'external_ip': iface_data['accessConfigs'][0]['natIP'],
+        }
+
+    def _blacklisted_by_tag(self, instance):
+        """Check if instance has any blacklisted tags.
```

Comment:
FYI no need to have docstrings for private methods :) if you want to have them (and I'd _prefer_ not to, but open to discussion) we should be consistent.

Reply:
I added them here as a way to give more context around what's happening because I found this opaque in zonextgen. If that's too much, I can rewrite these as comments.

Reply:
I'd hate to nitpick, but yeah I think I'd like these as regular comments...